swatch-2595: retry billable usage no billing account

kflahert opened the pull request:
<!-- Replace XXXX with the issue number. Issue will be auto-linked -->
Jira issue: SWATCH-2595

## Description
<!-- Provide a description of this PR.  Try to provide answers to "what", "how",
and "why" -->
Since we removed azure_tenant_id from the billing_account_id we now need to retry anytime a subscription is not found for a given billable usage. At first the subscription will have no billing_account_id and will eventually be updated to include one and retries should then succeed. 

## Testing
<!--
When possible, please use commands a developer can directly paste or modify
-->
IQE Test MR: <!-- IQE MR link here -->

### Setup
<!-- Add any steps required to set up the test case -->
1. Insert test data:
```
INSERT INTO public.offering VALUES ('BASILISK', 'BASILISK', NULL, 0, 0, 0, 0, 'BASLISK', 'Premium', 'Production', NULL, NULL, NULL, NULL, NULL);
INSERT INTO public.sku_product_tag VALUES ('BASILISK', 'BASILISK');
INSERT INTO public.subscription VALUES ('BASILISK', 'org123', '12345', 1, '2023-08-01 18:08:41.764447+00', '2025-09-21 15:06:29.374969+00', 'testing;testing;testing', '12222', 'azure', null);

INSERT INTO public.billable_usage_remittance VALUES (NULL, 'org123', 'BASILISK', 'Instance-hours', '2024-05', 'Premium', 'Production', 'azure', 'testing', 15, now(), '2024-05-31 02:10:28+00', NULL, 'AZURE', '4116be18-d614-437c-8c69-35244e2657ac', NULL, NULL, NULL);
INSERT INTO public.billable_usage_remittance VALUES (NULL, 'org123', 'BASILISK', 'Instance-hours', '2024-05', 'Premium', 'Production', 'azure', 'testing', 42, (now()::date + interval '6 hours'), '2024-05-31 02:10:28+00', NULL, 'AZURE', '64dc69e4-d083-49fc-9569-ebece1dd1408', NULL, NULL, NULL);
```

2. Start subscriptions, billable-usage, and azure-producer services:

`DEV_MODE=true ./gradlew :bootRun`

```
SERVER_PORT=8001 QUARKUS_MANAGEMENT_ENABLED=false SWATCH_INTERNAL_SUBSCRIPTION_ENDPOINT=http://localhost:8000 ./gradlew :swatch-producer-azure:quarkusDev
```

```
SERVER_PORT=8002 QUARKUS_MANAGEMENT_ENABLED=false KSTREAM_BILLABLE_USAGE_AGGREGATION_WINDOW_DURATION=3s KSTREAM_BILLABLE_USAGE_AGGREGATION_GRACE_DURATION=0s ./gradlew :swatch-billable-usage:quarkusDev
```

### Steps
<!-- Enter each step of the test below -->
4. Process retries to trigger billable usages being sent: 
`http POST :8002/api/swatch-billable-usage/internal/rpc/remittance/processRetries?as_of=2025-01-01T00%3A00Z Origin:console.redhat.com x-rh-swatch-psk:placeholder'`

5. Wait 3 seconds and flush aggregates:
`'http PUT :8002/api/swatch-billable-usage/internal/rpc/topics/flush Origin:console.redhat.com x-rh-swatch-psk:placeholder'`
''
6. Billable usage records should be updated with a retry_after one hour in the future because billing_account_id is unavaible:
`SELECT retry_after FROM public.billable_usage_remittance WHERE org_id='org123';`

7. Insert a usage past the azure window of 96hrs:
`INSERT INTO public.billable_usage_remittance VALUES (NULL, 'org123', 'BASILISK', 'Instance-hours', '2024-05', 'Premium', 'Production', 'azure', 'testing', 42, '2024-05-01 02:10:28+00', '2024-05-31 02:10:28+00', NULL, 'AZURE', 'cf0fadee-2d07-421d-9cff-bca10adc2908', NULL, NULL, NULL);`

8. Repeat steps 3 and 4

9. Check logs swatch-producer-azure logs for:

`Skipping billable usage with id=cf0fadee-2d07-421d-9cff-bca10adc2908 orgId=org123 because the subscription was not found and the snapshot '2024-05-01T02:10Z' is past the azure time window of 'PT96H'`

Sgitario reviewed (changes requested):
Apart from some suggestions, the usage question and the way to detect the API returned 404, the testing instructions didn't work for me.

At first, these two statements didn't work for me:
```
INSERT INTO public.billable_usage_remittance VALUES (NULL, 'org123', 'BASILISK', 'Instance-hours', '2024-05', 'Premium', 'Production', 'azure', 'testing', 15, CURRENT_TIME, '2024-05-31 02:10:28+00', NULL, 'AZURE', '4116be18-d614-437c-8c69-35244e2657ac', NULL, NULL, NULL);
INSERT INTO public.billable_usage_remittance VALUES (NULL, 'org123', 'BASILISK', 'Instance-hours', '2024-05', 'Premium', 'Production', 'azure', 'testing', 42, CURRENT_TIME, '2024-05-31 02:10:28+00', NULL, 'AZURE', '64dc69e4-d083-49fc-9569-ebece1dd1408', NULL, NULL, NULL);
```

Because "CURRENT_TIME" is not recognised as a valid timestamp. Changing it to "now()" worked. 
But then, the constraint "billable_usage_remittance_unique" is raised because the above two statements are really the same, but with different values. Using "Cores" as the metric ID in the second worked. 

When I moved forward, the step 5 didn't work (the retry_after was never populated). 

Checking the tally and producer-azure logs, I see some traces saying "Skipping invalid billable usage " that seem related.

kflahert commented on swatch-producer-azure/src/main/java/com/redhat/swatch/azure/service/BillableUsageAggregateConsumer.java:
```diff
@@ -229,7 +230,7 @@ public AzureUsageContext lookupAzureUsageContext(BillableUsageAggregate billable
           billableUsageAggregate.getAggregateKey().getOrgId(),
           billableUsageAggregate.getAggregateKey().getSla(),
-          billableUsageAggregate.getAggregateKey().getMetricId(),
+          billableUsageAggregate.getAggregateKey().getUsage(),
```
We have a comment that it should be added eventually so I figured we should just leave it in: https://github.com/RedHatInsights/rhsm-subscriptions/blob/879f9fa15fe397e39461f394b02cd136c9c94c3d/src/main/java/org/candlepin/subscriptions/subscription/SubscriptionSyncController.java#L646

Sgitario replied:
+1 it makes sense

kflahert commented:
@Sgitario sorry for the bad test steps. I just updated them to fix the subscription to remove the billing_account_id(which should cause the retry error) and fixed the dates to use now() and now()-6hours. They should work now.

Sgitario commented on swatch-producer-azure/src/main/java/com/redhat/swatch/azure/service/BillableUsageAggregateConsumer.java:
```diff
@@ -240,10 +242,13 @@ public AzureUsageContext lookupAzureUsageContext(BillableUsageAggregate billable
+                    error ->
+                        (Integer.toString(Status.NOT_FOUND.getStatusCode()))
+                            .equals(error.getStatus()));
+        if (isNotFound) {
```
I've checked that doing `Status.NOT_FOUND.getStatusCode() == Status.NOT_FOUND.getStae.getResponse().getStatus()` would also work (don't need to parse the integer to string and also loop over the list of errors. 

This is a minor suggestion, feel free to change it or not.

Sgitario reviewed (approved):
Now, everything works as stated in the instructions. 
I only added a minor suggestion about the code, but not blocker.

diegomaranhao commented:
/retest

diegomaranhao commented:
/retest
